### PR TITLE
Pocket node entrypoint, module interface and application specific bus 

### PR DESCRIPTION
**Pull request by Olshansk**
# Context

This is the first of several PRs to address https://github.com/pokt-network/pocket/issues/21.

A lot of the code here was copied over from the `prototype` in https://github.com/pokt-network/pocket/pull/27, with some fixes, cleanup, modularization. 

**IMPORTANT NOTE TO THE REVIEWER**: Note that some code was deleted so just a skeleton is submitted and should be added back when the module-specific code is added back in.

## Details

- Various additions to the .gitignore file
- Makefile addition: temporary / experimental targets to enable V1 prototype development
- Add Dockerfiles for validator, debug client and Protobufs generation
- Add a 1st iteration version of the configuration file
- Add a 1st iteration version of the genesis file
- Add a docker-compose file for 4 validators and a debug client
- Add a debug client with a placeholder trigger
- Skeleton implementations of all 4 modules
- Update all go dependencies (go.mod, go.sum, etc…)
- Migrated first iteration of crypto library from pre-prototype to the prototype

## Next Steps
* Submit this PR as the foundation for pocket V1
* Prepare a small "design document" (< 1 page) on the structure of the bus and modules
* Write up a small document on how to use these (for contributors and new employees)

## TODOs as a result of the discussion PR
** (will create an issue and add these before submission)
- [ ] **[Small]** Remove makefile altogether
- [ ] **[Small]** Remove protobuf generation from host machine to Dockerfiles
- [ ] **[Small]** Add documentation for Goland & DLV debugging
- [ ] **[Medium]** Migrate all useful targets from Makefile to mage
- [ ] **[Large]** Discuss & design how `LoadConfig` should be implemented
- [ ] **[Ongoing]** Iterate on and define the structure of the configuration file
- [ ] **[Ongoing]** Iterate on and define the structure of the genesis file

## 1. Pull changes (if not already local):
```
$ git clone git@github.com:pokt-network/pocket.git pocket
$ cd pocket
$ git checkout -b issue21/prototype/entrypoint origin/issue21/prototype/entrypoint
```

## [Optional] 2. View available commands:
```
$ make
```

## 3. Clean environment:
```
$ make docker_wipe # Press `y` if prompted
```

### 3.1 Shell 1:
```
$ make compose_and_watch
```

### 3.2 Shell 2:
```
$ make client_start
$ make client_connect
```

### 3.3. Run commands

Inspect that a placeholder events gets continually sent to the main bus

## [Optional] 4. Try running the other `make` targets:
```
$ make mockgen
$ make test_all
$ make protogen_local
```

### CODE REVIEW

File: consensus/module.go
Comment (andrewnguyen22) [inline consensus/module.go:11]: Remove unused constant

File: consensus/module.go
Comment (andrewnguyen22) [inline consensus/module.go:54]: I don't understand why transaction needs to be different than message? Isn't a transaction just a message?

File: p2p/module.go
Comment (andrewnguyen22) [inline p2p/module.go:14]: Let's try to leave out the prefix `pocket` from things when possible

File: p2p/module.go
Comment (andrewnguyen22) [inline p2p/module.go:32]: I see p2p and I see network

We need terminology to be completely unified before moving forward

File: pre_p2p/module.go
Comment (andrewnguyen22) [inline pre_p2p/module.go:1]: Justification for why this should be a TLD and not within P2P?

File: prototype/Makefile
Comment (andrewnguyen22) [inline prototype/Makefile:23]: Can the prototype dir be historical only? I would prefer to keep it as a branch / commit but lose it on main merge

File: shared/config/config.go
Comment (andrewnguyen22) [inline shared/config/config.go:12]: Is this planning on being used? If not - let's remove

File: prototype/PROTOTYPE.md
Comment (andrewnguyen22) [inline prototype/PROTOTYPE.md:27]: Doesn't exist

File: shared/modules/utility_module.go
Comment (andrewnguyen22) [inline shared/modules/utility_module.go:11]: TODO: There's got to be a better way to structure this; shouldn't block this PR, but note to self

File: shared/types/events.go
Comment (andrewnguyen22) [inline shared/types/events.go:14]: Let's get the format down, weird that some are all caps and others are all lower... Surely persistence isn't a private type

File: shared/handler.go
Comment (andrewnguyen22) [inline shared/handler.go:15]: See earlier comment about transaction vs message

File: shared/node.go
Comment (andrewnguyen22) [inline shared/node.go:20]: Just wanted to confirm that we wanted to do inheritance here and not interface enforcement...

Comment (andrewnguyen22) [reply shared/node.go:20]:
```
Node{
		Module:    nil, // doesn't need to be here...
		pocketBus: nil,
		Address:   "",
	}
```

File: utility/module.go
Comment (andrewnguyen22) [inline utility/module.go:15]: I'd prefer if you made this param explicitly not used

`func Create(_ *config.Config) (modules.UtilityModule, error) {`

**Comment by andrewnguyen22** [comment]
I don't think this was THIS PR - but `gofmt` is complaining for mage.go
Considering your adding the contribution guide here, might be appropriate to fix

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: consensus/module.go
Comment (Olshansk) [reply consensus/module.go:11]: Removed.

File: consensus/module.go
Comment (Olshansk) [reply consensus/module.go:54]: tl;dr Removing HandleTransaction for now.

Good point. I was thinking of a `Message` as being consensus specific and `Transaction` as being more at the utility level. However, these can both be wrapped in an envelope that'll be opened by the consensus module and we can always add it back later if need be.

File: p2p/module.go
Comment (Olshansk) [reply p2p/module.go:14]: 👍  Replaced all instances of `pocketBus` with `bus`.

File: p2p/module.go
Comment (Olshansk) [reply p2p/module.go:32]: Done. Using `p2p` everywhere now.

File: pre_p2p/module.go
Comment (Olshansk) [reply pre_p2p/module.go:1]: Added this comment to the top of the file:

```
// TODO(team): This is a Top Level Module since it is a temporary parallel to the
// real `p2p` module. It should be removed once the real `p2p` module is ready but
// is meant to be a "real" replacement for now.
```

👍 ?

File: prototype/Makefile
Comment (Olshansk) [reply prototype/Makefile:23]: I merge it to the v1-milestone branch in the other PR for history/reference purposes.

I expect us to have a PR where we just delete it in the future (once main is clean and functioning), and that ways we'll have the historical context via git.

File: shared/config/config.go
Comment (Olshansk) [reply shared/config/config.go:12]: tl;dr removed the constant but keeping the file for now.

It's still being used in the config files and needed for the consensus prototype, but the constant was not being used.

File: shared/handler.go
Comment (Olshansk) [reply shared/handler.go:15]: Removed.

File: utility/module.go
Comment (Olshansk) [reply utility/module.go:15]: Done. Updated persistence module as well.

File: shared/node.go
Comment (Olshansk) [reply shared/node.go:20]: Done.

File: shared/modules/utility_module.go
Comment (Olshansk) [reply shared/modules/utility_module.go:11]: Added TODO in the code.

**Comment by Olshansk** [comment]
@andrewnguyen22 Addressed everything except for a couple outstanding comments. 

Re formatting, I added `make gofmt` target but it didn't seem to update much.

File: pre_p2p/module.go
Comment (andrewnguyen22) [reply pre_p2p/module.go]: But I feel as if it might be useful to keep a minimal mock of P2P indefinitely... The reason being, testing and simplicity. I believe all things P2P should live under the P2P dir.

File: shared/types/events.go
Comment (andrewnguyen22) [reply shared/types/events.go:14]: @Olshansk?

Comment (andrewnguyen22) [reply shared/types/events.go:14]: To be clear, my recommendation is:
```
CONSENSUS_MODULE SourceModule = "consensus"
P2P              SourceModule = "p2p"
PERSISTENCE      SourceModule = "persistence" // notice the name of the var
```

File: pre_p2p/module.go
Comment (Olshansk) [reply pre_p2p/module.go]: tl;dr Moved `pre2p` into `p2p`.

- I don't really see `pre2p` as a mock as much as it is an alternative implementation. 
- We'll also have REAL mocks with `mockgen` (or w/e we decide to use for mocks) 
- Ideally, the real P2P module should be able to replace pre2p 100%
- I don't feel too strongly whether it's a TLD or not, so moving into p2p

File: prototype/PROTOTYPE.md
Comment (Olshansk) [reply prototype/PROTOTYPE.md:27]: Reverted.

File: shared/types/events.go
Comment (Olshansk) [reply shared/types/events.go:14]: Done. I also deleted the majority of topics in this file and we can add them back on a at-need basis.

File: shared/pocket_bus.go
Comment (iajrz) [inline shared/pocket_bus.go:21]: this is a confusing comment.

File: shared/types/events.go
Comment (iajrz) [inline shared/types/events.go]: CamelCase constant names are idiomatic.

**Review by iajrz — CHANGES_REQUESTED** [review]
This PR is bigger than I'd've expected; there are three particular things that we should definitely align on to shape the code:
- makefile vs magefile
- configuration loading
- version flag / version information handling

File: Makefile
Comment (Olshansk) [reply Makefile:24]: Done.

File: .gitignore
Comment (Olshansk) [reply .gitignore:13]: Done.

File: Makefile
Comment (Olshansk) [reply Makefile:1]: I added a TODO at the top of the `Makefile`. 

I don't want this discussion (which is not a small discussion) to block this PR since it's a bit out of scope of what this tried to do, but we could discuss this in another Github issue or protocol hour instead.

File: Makefile
Comment (Olshansk) [reply Makefile:91]: I renamed it to `protogen_docker` but note the TODO I left for Hamza since it wasn't working for me and is also outside the scope of this PR.

I'm sticking with `protogen_local` to unblock the app-specific bus implementation.

File: cmd/pocket/main.go
Comment (Olshansk) [reply cmd/pocket/main.go:11]: Adding a TODO to update documentation/explanation instead.

File: go.mod
Comment (Olshansk) [reply go.mod:1]: Updated everywhere.

File: shared/config/config.go
Comment (Olshansk) [reply shared/config/config.go:18]: Thank. Using `pre2p` everywhere now.

File: shared/pocket_bus.go
Comment (Olshansk) [reply shared/pocket_bus.go:21]: Comment removed.

File: shared/handler.go
Comment (Olshansk) [reply shared/handler.go:14]: Prefacing the log with `WARN` and adding a comment to discuss if it should panic, but do not think it is a blocker for this PR.

File: shared/config/config.go
Comment (Olshansk) [reply shared/config/config.go:50]: Good point,  but I think this discussion exceeds the scope of this PR and might be better suited for an issue related to "configuration & parameters" rather than the "app-specific skeleton bus".

File: cmd/pocket/main.go
Comment (Olshansk) [reply cmd/pocket/main.go:20]: The flag is of type `bool` so %t had a warning with it. I just fixed the warning but adding a TODO since fixing this is outside the scope of this PR.

File: shared/types/events.go
Comment (Olshansk) [reply shared/types/events.go]: Converted to CamelCase.

File: shared/crypto/sha3.go
Comment (Olshansk) [reply shared/crypto/sha3.go:14]: Done.

File: shared/crypto/keys.go
Comment (Olshansk) [reply shared/crypto/keys.go:10]: Adding TODO. Trying to prevent this PR from getting too out of scope.

File: shared/config/config_test.go
Comment (Olshansk) [reply shared/config/config_test.go:10]: IMO config loading tests are a low priority while we still don't have a well-defined definition/list of configurations. Adding a TODO for now.

**Comment by Olshansk** [comment]
@iajrz Addressed some (but not all) of your comments. 

I think there is A LOT of work we need around configuration testing/loading/validation, but I'm trying to find a balance to avoid getting too hung up on this PR, which is targeted at the app-specific bus along with minimal configuration/tooling to enable it.

If we need to create other issues, or discuss, please let me know.

**Review by andrewnguyen22 — APPROVED** [review]

File: Makefile
Comment (iajrz) [reply Makefile:3]: I am hard pressed to see code added in this PR as out of scope for this PR.

File: cmd/pocket/main.go
Comment (iajrz) [reply cmd/pocket/main.go]: It's mentioned in [the dependency docs](https://github.com/pokt-network/pocket/tree/main/docs/deps) where the dependency in mage is justified; it's also related to the conversation about `ldflags` we had on #14 so I thought it was understood. Where should the explanation be found?

**Review by iajrz — APPROVED** [review]

**Comment by derrandz** [comment]
Github projects should really prompt the user with "Do you want to save the changes" so that the title does not change as quick.

**Pull request merged**
